### PR TITLE
Update `macOS` testing strategy to run on latest `arm` architecture

### PR DESCRIPTION
**Pull request by user27182**
### Overview

Add testing with `macos-latest`.

### Details
macOS was pinned to `macos-12` in https://github.com/pyvista/pyvista/pull/5960

However, this means that we are not testing for compatibility with newer macbooks which use ARM processors instead of intel processors. The new architecture breaks compatibility with some of pyvista's optional dependencies, e.g. `embreex`, but this is undetected without testing on the newer architecture.

### CODE REVIEW

File: .github/workflows/testing-and-deployment.yml [hunk 1]
```diff
@@ -21,21 +21,22 @@ env:
...
-    runs-on: macos-12
     name: Mac OS Unit Testing
     strategy:
       fail-fast: false
       matrix:
-        python-version: ["3.9", "3.11"]
+        config:
+          - { os: macos-12, python-version: 3.9 } # intel processor
+          - { os: macos-latest, python-version: 3.12 } # arm processor
+    runs-on: ${{ matrix.config.os }}
```
Comment (user27182) [inline .github/workflows/testing-and-deployment.yml:31]: Not sure if there's a reason we only have two configs in the matrix, should we add more? E.g. windows testing has a runner for every supported python version
```suggestion
        config:
          - { os: macos-12, python-version: 3.9 } # intel processor
          - { os: macos-13, python-version: 3.10 } # intel processor
          - { os: macos-14, python-version: 3.11 } # arm processor
          - { os: macos-latest, python-version: 3.12 } # arm processor
    runs-on: ${{ matrix.config.os }}
```

**Review by banesullivan — APPROVED** [review]
This is excellent and well overdue! Thanks for updating these and setting up the matrix from cross platform testing on Mac

**Comment by banesullivan** [comment]
It appears I may need to edit the "required checks" for CI

**Review by tkoyama010 — APPROVED** [review]

**Comment by tkoyama010** [comment]
I added "macos-*" requirements

**Pull request merged**